### ISSUE TITLE
Tutorial says FK is not needed, is a orphan object ok?

### ISSUE THREAD

**Issue by pluppplupp** [Starcounter/InvoiceDemo#46]
![2015-12-14_12-50-08](https://cloud.githubusercontent.com/assets/1765023/11780319/68413da6-a261-11e5-8dc3-00d766ed5218.png)

This gives me some unanswered questions. Can I remove the Invoice relating to the InvoiceRow? Will the InvoiceRow be orphaned?

**Comment by miyconst** [comment]
Yes, this is by design.

If you want to make sure that all `InvoiceRow` being deleted with the `Invoice`, then you have two options:

#### Implement `IEntity` interface in `Invoice` class:

```cs
[Database]
public class Invoice: IEntity {
    public void OnDelete() {
        //Delete the rows here
    }
}
```

#### Create a commit hook for `Invoice` class and delete its rows in there:

```cs
Hook<Invoice>.BeforeDelete += (sender, entity) => {
    //Delete the rows here
};
```

**Comment by pluppplupp** [comment]
And if I want classic referential integrity check I would throw an exception?

**Comment by miyconst** [comment]
Yes, exactly.

**Comment by pluppplupp** [comment]
Mabye, we should explain this in the tutorial or reference a "read more" page?!

**Comment by miyconst** [comment]
For sure.

**Comment by miyconst** [comment]
@warpech can you take this one?

**Comment by warpech** [comment]
Duplicate of: https://github.com/Starcounter/Starcounter/issues/3035

Thanks a lot @pluppplupp for describing the issue! 

I think that a new page "Cascading deletes" should be added to http://starcounter.io/guides/database/ . I am not the best person to write it, but I can surely copy and paste the content from here  and https://github.com/Starcounter/Starcounter/issues/3035.

@per-samuelsson do you agree?

**Comment by per-samuelsson** [comment]
> @per-samuelsson do you agree?

Yes and no. Yes, because we need to document this. No, because I'm not satisfied with how the current API look.

But maybe its one of those "we're not going to have time to refine this API; lets just document what is, and see if we can improve in the next major".

If we decide to go that last way, we need to choose what model to promote I guess. Documenting both will just introduce confusion, especially if we dont clearly express pros/cons of each alternative.

**Comment by warpech** [comment]
Solution idea: Add a sidenote to this page that links to http://starcounter.io/guides/database/object-identity/

This page shows how to add a sidenote: http://starcounter.io/style-guide/

**Comment by dan31** [comment]
I gave it some thinking today, my current opinion is that optional/ad-hoc referential integrity, as we have today in Starcounter, corresponds to "you don't pay for what you don't use" principle. Taking a look at referential integrity in general, Imagining Starcounter classes be ordinary .NET or even C++ classes. For them, one should implement some form of reference counting (GC in .NET or shared_ptr in C++) or, operating in a subject domain, know what happens and decide who is an object owner, so that when object owner is removed the owned objects are removed as well. That is what I mean with the latter. In your ```InvoiceRow``` example, you know that ```InvoiceRow``` belongs to ```Invoice```, so that when you remove ```Invoice```, all ```InvoiceRows``` are a matter of deletion. However, if you model a graph, this will differ: removing the edge doesn't remove the vertices adjacent, but removing the vertex you'd probably want to perform also with removing the edges adjacent (but not necesserily). In my opinion it is totally fine as of today to explain practices simplifying adding referential integrity and cascaded deletes to Starcounter.
